edx-solutions/xblock-drag-and-drop-v2
# [MCKIN-5774] Wait 500 milliseconds before starting drag on touch.

Wait 500 milliseconds before starting drag on touch.

This allows users to scroll the item bank with their finger without accidentally starting to drag items.

Drag events initiated via mouse still start without a delay, since scrolling works differently on desktops.

**JIRA tickets**: OC-3059, MCKIN-5774

**Dependencies**: None

**Testing instructions**:

1. Add the Drag and Drop demo course to your course in various configurations.
2. Test mouse dragging. It should behave just like before this patch.
3. Test touch dragging. Touch dragging should only start 500ms after you hold the finger on the item without moving. If you put your finger on the item, but move the finger before 500ms pass, the drag should not start and default browser behaviour should kick in (i.e. page scrolling).
4. Make sure to test in both Standard and Assessment modes. In assessment mode, items can be moved between zones after they are placed. They can also be dragged back to the item bank.

**Author notes and concerns**:

1. I found no way to implement this with jQuery UI draggable, so I had to re-implement drag and drop interactions, which in some ways makes things cleaner because it doesn't bypass virtualDOM.
2. In assessment mode, when you drag an item that was already placed in one of the zones, at the moment the item is picked up, the dimensions of the item may change (and are equal to item's dimensions in the bank). It would be possible to prevent that, but would take more time than I was able to spend on this ticket. It would be a nice future improvement.
3. The existing tests all pass, but I am not sure how to add tests for touch interactions, since FF driver for Selenium doesn't support them. Suggestions welcome.

**Reviewers**
- [ ] @bradmerlin
